### PR TITLE
Disable block selection when resizing image

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR fixes the problem reported in https://github.com/WordPress/gutenberg/issues/3362. It adds a functionality to disable block selection and then disables block selection during the image resize.

## How Has This Been Tested?

1. Add an image block (with an image) and some paragraphs below it.
2. Resize it to small size.
3. Resize it back up enthusiastically, expecting the grabber to hit the corner of the block, so you know it's as big as it can be.
4. Verify no block selection happened.
5. Try to select blocks as before and verify you are still able to select them (when not resizing an image).

### CODE REVIEW

File: blocks/library/image/block.js
Comment (youknowriad) [inline blocks/library/image/block.js]: While this works, It raises the question of whether a block can subscribe to the state or not? We never didi this before and I think if we were to do it, it should be via the same mechanisms exposing the state for extensions. cc @aduth 

Do you think we can find a way to fix this bug in a generic way for all blocks?

Comment (jorgefilipecosta) [reply blocks/library/image/block.js]: I updated this PR, now we don't access the store in the blocks. And we allow any block to disable/enable selection. I think when blocks implement a mouse movement equal to the one used for multi-selection they will need to disable multi-selection. I'm not seeing another possibility. The set of movements, in my opinion, is limited it will probably be resizing and drag and drop operations. And now we are proving a mechanism for blocks to solve this problem.

File: editor/components/block-list/block.js
Comment (mcsf) [inline editor/components/block-list/block.js]: Here and across the diff, should read _isSelectionDisabled_ instead. The term _toggleSelection_ is fine, however.

File: editor/actions.js
Comment (mcsf) [inline editor/actions.js]: I'd go for just _toggle selection_ (in both the action name and action type). Maybe even _toggle block selection_.

**Review by mcsf — COMMENTED** [review]
Is an interface based around state inspection + state toggling adequate? Looking at `blocks/library/image`, I'd have expected a simpler mechanism where the Image block unidirectionally informs the framework:

```js
onResizeStart={ ( … ) => {
  disableSelection(); // or even "grabSelection"
} }
onResizeStop={ ( … ) => {
  enableSelection(); // or even "releaseSelection"
} }
```

File: blocks/library/image/block.js
Comment (aduth) [inline blocks/library/image/block.js]: Minor: I think we should rename this prop to either `disableSelection` or `selectionDisabled` to read better.

File: blocks/library/image/block.js
Comment (aduth) [inline blocks/library/image/block.js]: Maybe the usage should be to specify the value we want in toggling to avoid the need for the condition? `toggleSelection( false );`. Could it be clearer if we're consistent with what we're allowing here? A bit confusing to work with a disabled prop to toggle it as enabled; wondering if it should be `selectionDisabled`, `toggleSelectionDisabled`, or the inverse: `selectionEnabled`, `toggleSelection`.

**Review by aduth — COMMENTED** [review]
I agree with previously noted concerns that it's not great that a block implementer would need to worry themselves with selection behavior of the editor. It speaks more to a problem of how we implement multi-selection, and at least I could see the need for tackling this to be quite limited to cases where click-dragging occurs.

**Comment by jorgefilipecosta** [comment]
I address the code points raised in the reviews, thank you @aduth, @mcsf, and @youknowriad. I agree the selection behavior is not something block developers should worry about, and this solution is far from perfect.
I would say is a simple solution that works for now. Maybe one day the selection mechanism changes or we find a better way. As noted this problem only affects blocks with "click-drag" so most blocks don't have to worry about that for only the image block makes uses of it :)

**Comment by youknowriad** [comment]
@jorgefilipecosta Thanks for handling and I agree with you, it's not something easy to do without adding a block API (impossible right now?). So, yeah I think this is good to go. Might be good to have another opinion before merging (cc @aduth @mcsf )

File: editor/actions.js
Comment (aduth) [inline editor/actions.js]: Can we document this function?

File: editor/reducer.js
Comment (aduth) [inline editor/reducer.js]: Minor: "selection" here seems redundant, i.e. `blockSelection.isSelectionEnabled` vs. `blockSelection.isEnabled`

File: editor/test/reducer.js
Comment (aduth) [inline editor/test/reducer.js]: These lines are quite long and I'd suggest considering to split them on multiple lines.

**Review by aduth — APPROVED** [review]
While not ideal, I think I'm okay with a block which knows that it needs to override selection behavior via its own dragging behavior to toggle selection. 👍 

I found that there are errors logged when resizing images, and that Redux DevTools becomes very slow when doing so (signalling some large payloads or frequent dispatches). This doesn't appear to be specific to the changes here though.

**Review by mcsf — APPROVED** [review]
I second Andrew's final suggestions; other than, LGTM!

**Comment by jorgefilipecosta** [comment]
Thank you for your reviews @mcsf and @aduth, I applied the suggestions before the merge.

File: editor/selectors.js
Comment (aduth) [inline editor/selectors.js]: As having trouble remembering what it means to have "selected enabled", given the various types of selections which can occur in Gutenberg, I found this comment insufficient in explaining what I can expect as the meaning of the value returned by this selector.

**Pull request merged**
